src/Common/src/System/Marvin.cs

```diff
@@ -14,74 +14,52 @@ internal static class Marvin
         /// Convenience method to compute a Marvin hash and collapse it into a 32-bit hash.
         /// </summary>
         [MethodImpl(MethodImplOptions.AggressiveInlining)]
-        public static int ComputeHash32(ref byte data, int count, ulong seed)
+        public static int ComputeHash32(ReadOnlySpan<byte> data, ulong seed)
         {
-            long hash64 = ComputeHash(ref data, count, seed);
+            long hash64 = ComputeHash(data, seed);
             return ((int)(hash64 >> 32)) ^ (int)hash64;
         }
 
         /// <summary>
         /// Computes a 64-hash using the Marvin algorithm.
         /// </summary>
-        public static long ComputeHash(ref byte data, int count, ulong seed)
+        public static long ComputeHash(ReadOnlySpan<byte> data, ulong seed)
         {
-            uint ucount = (uint)count;
             uint p0 = (uint)seed;
             uint p1 = (uint)(seed >> 32);
 
-            int byteOffset = 0;  // declared as signed int so we don't have to cast everywhere (it's passed to Unsafe.Add() and used for nothing else.)
-
-            while (ucount >= 8)
+            if (data.Length >= sizeof(uint))
             {
-                p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset));
-                Block(ref p0, ref p1);
+                ReadOnlySpan<uint> uData = data.NonPortableCast<byte, uint>();
 
-                p0 += Unsafe.As<byte, uint>(ref Unsafe.Add(ref data, byteOffset + 4));
-                Block(ref p0, ref p1);
+                for (int i = 0; i < uData.Length; i++)
+                {
+                    p0 += uData[i];
+                    Block(ref p0, ref p1);
+                }
 
-                byteOffset += 8;
-                ucount -= 8;
+                // byteOffset = data.Length - data.Length % 4
+                // is equivalent to clearing last 2 bits of length
+                int byteOffset = data.Length & (~3);
```

Comment:
I found that using `byteOffset = data.Length - data.Length % 4` directly gives a perf hit of at least 5%

Reply:
Good comment, please add it to the code.

Reply:
done